core/trino-main/src/main/java/io/trino/sql/planner/optimizations/PropertyDerivations.java

```diff
@@ -580,6 +580,14 @@ public ActualProperties visitExchange(ExchangeNode node, List<ActualProperties>
             // are computed for local exchanges.
             // TODO: implement full properties for local exchanges
             if (node.getScope() == LOCAL) {
+                if (inputProperties.size() == 1) {
+                    ActualProperties inputProperty = inputProperties.get(0);
+                    if (inputProperty.isEffectivelySingleStream()) {
```

Comment:
any way to test it?

Reply:
It would get tested as part of https://github.com/trinodb/trino/pull/6634, but I don't see any straightforward way to specifically test it here

Reply:
What if `node` is `REPARTITION`, and the source has `GroupingProperty`?

Reply:
I thought that if we have a single stream with grouping then a repartition of that should result in multiple streams with each stream retaining grouping property. Let me know if that's correct or the grouping property will be broken in this scenario ?

Reply:
I'm not sure I understand it correctly (AddExchanges and PropertyDerivations terribly lack documentation...), but I imagine that repartitioning can override the input grouping.

Also, I am concerned about the case when both `node` and input have SortingProperties. Here you concatenate the properties lists: first parent's, then child's. If duplicates appear, are they removed later? And if the sorting properties are different, is it correct to keep child's properties? (in other words, is the sort stable?)

Reply:
I made a change to ignore child properties in case `node` has a ordering scheme